### PR TITLE
Implement softmax_backward

### PR DESCRIPTION
**Pull request by vtsilytskyiTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/27796

### Problem description
Lack of softmax_backward op.

### What's changed
Basic implementation of softmax_backward fused operation with the unit test.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_device_operation.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_device_operation.cpp]: TT_FATAL

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.cpp]: it's not clear why you need reshape

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp]: Querying runtime args can be quite inefficient, consider optimizing this code this way
```
auto reader_runtime_args = GetRuntimeArgs(program, reader_kernel_id);
for (....) {
   auto& reader_runtime_args_per_core = reader_runtime_args[core.x][core.y];
   .... 
}
```

Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp:112]: that's a huge amount of memory needed even for the case when it clearly fits to L1 memory. 
For 8192 -> it will require 2MB of L1 cache, at least output buffer might not require full intermediate dimension 

It can be the case that pushing/waiting memory block by block for reader/writer be more efficient with reasonable overlap

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp]: this needs to correctly process case when logical shape is not equal to padded shape (reduction result might not be correct)

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp
Comment (rfurko-tt) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp:112]: we also do not utilize idea of double buffering. We can't read next batch of data in reader while we are computing current batch of data

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp:64]: please rename cb handles in a proper way. `y_cb_id`, `grad_cb_id`, `grad_minus_sum_cb_id` etc
it would be so much easier to read the code

File: tests/ttnn/unit_tests/operations/test_softmax_backward.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+from numpy import absolute
+import torch
+import pytest
+import ttnn
+from tests.ttnn.unit_tests.operations.eltwise.backward.utility_funcs import data_gen_with_range_dtype
+
+
+def torch_to_device(pt_tensor: torch.Tensor, ttnn_dtype: ttnn.types, device: ttnn.Device) -> ttnn.Tensor:
+    return ttnn.Tensor(pt_tensor, ttnn_dtype).to(ttnn.TILE_LAYOUT).to(device)
```
Comment (rfurko-tt) [inline tests/ttnn/unit_tests/operations/test_softmax_backward.py]: there is a function ttnn.from_torch

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp]: pass this defines in program factory

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/dataflow/reader_softmax_backward.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/dataflow/reader_softmax_backward.cpp]: why don't you use common method? 
```
noc_async_read_tile
```

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.hpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.hpp]: we allow dim to be -1, but it's unsigned int. Please change underlying type to int32_t

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp:64]: Done

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/dataflow/reader_softmax_backward.cpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/dataflow/reader_softmax_backward.cpp]: Good point! As I see some of APIs is deprecated. The best way to read is `noc_async_read_page`, which I'm going to use.

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_device_operation.cpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_device_operation.cpp]: Done

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp]: Done.

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.cpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.cpp]: Indeed, it is redundant. Removed.

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    uint32_t src1_tile_idx) {
+    // Step 1: Compute grad - sum(y * grad) and store in intermediate CB
+    // Note: caller ensures src1, sum_reduce_cb have required tiles available
+    // sum_reduce_cb always has the scalar at index 0
+    cb_reserve_back(intermed_cb_id, 1);
+
+    sub_tiles_init(src1_cb_id, sum_reduce_cb_id);
+
+    tile_regs_acquire();
+    sub_tiles(src1_cb_id, sum_reduce_cb_id, src1_tile_idx, 0, 0);  // grad[w] - sum[0]
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp]: I think this is an elementwise sub, which won't work because sum_reduce_cb_id has only the first column populated properly. You should look into using a broadcast sub.

File: tests/ttnn/unit_tests/operations/test_softmax_backward.py (hunk 1 above)
Comment (vtsilytskyiTT) [reply tests/ttnn/unit_tests/operations/test_softmax_backward.py]: Yep, good alternative to my function, thanks!

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.hpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.hpp]: Done

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp (hunk 2 above)
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp]: Yes, indeed it was a wrong sub. Now works well, thank you.

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp]: Done

**Comment by vtsilytskyiTT** [comment]
Closing in favor of https://github.com/tenstorrent/tt-metal/pull/31580, production-ready implementation.

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp:112]: Implemented streaming algo, which handles any tensor length. See https://github.com/tenstorrent/tt-metal/pull/31580

File: ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp
Comment (vtsilytskyiTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp]: Done in https://github.com/tenstorrent/tt-metal/pull/31580
